crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,365 @@
+            }
+            Int(index) => {
+                let index = self.convert_py_index(index, IndexFor::Lookup)?;
+                let value: ElementType = value.extract()?;
+                let mut cached_entry = self.get_or_cache(py, value)?;
+                swap(&mut cached_entry, &mut self.data[index]);
+                self.drop_from_cache(py, cached_entry)
+            }
+        }
+    }
+
+    pub fn insert(&mut self, py: Python<'_>, index: isize, value: ElementType) -> PyResult<()> {
+        let index = self.convert_py_index(index, IndexFor::Insertion)?;
+        let cache_entry = self.get_or_cache(py, value)?;
+        self.data.insert(index, cache_entry);
+        Ok(())
+    }
+
+    pub fn pop(&mut self, py: Python<'_>, index: Option<isize>) -> PyResult<PyObject> {
+        let index = index.unwrap_or(max(0, self.data.len() as isize - 1));
+        let item = self.__getitem__(py, Int(index))?;
+        self.__delitem__(py, Int(index))?;
+        return Ok(item);
+    }
+
+    pub fn append(&mut self, py: Python<'_>, value: ElementType) -> PyResult<()> {
+        self.insert(py, self.data.len() as isize, value)
+    }
+
+    pub fn extend(&mut self, py: Python<'_>, itr: Vec<ElementType>) -> PyResult<()> {
+        for v in itr {
+            self.append(py, v)?;
+        }
+        Ok(())
+    }
+
+    pub fn clear(&mut self, py: Python<'_>) -> PyResult<()> {
+        let mut to_drop = vec![];
+        swap(&mut self.data, &mut to_drop);
+        for entry in to_drop.into_iter() {
+            self.drop_from_cache(py, entry)?;
+        }
+        Ok(())
+    }
+
+    #[classattr]
+    const __hash__: Option<Py<PyAny>> = None;
+
+    fn __richcmp__(
+        slf: &PyCell<Self>,
+        other: &PyAny,
+        op: CompareOp,
+        py: Python<'_>,
+    ) -> PyResult<PyObject> {
+        match op {
+            CompareOp::Eq => CircuitData::equals(slf, other).map(|r| r.into_py(py)),
+            CompareOp::Ne => CircuitData::equals(slf, other).map(|r| (!r).into_py(py)),
+            _ => Ok(py.NotImplemented()),
+        }
+    }
+}
+
+enum IndexFor {
+    Lookup,
+    Insertion,
+}
+
+impl CircuitData {
+    fn convert_py_slice(&self, py: Python<'_>, slice: &PySlice) -> PyResult<Vec<isize>> {
+        let dict: HashMap<&str, PyObject> =
+            HashMap::from([("s", slice.into()), ("length", self.data.len().into_py(py))]);
+        py.eval(
+            "list(range(*s.indices(length)))",
+            None,
+            Some(dict.into_py_dict(py)),
+        )?
+        .extract()
+    }
+
+    fn convert_py_index(&self, index: isize, kind: IndexFor) -> PyResult<usize> {
+        let index = if index < 0 {
+            index + self.data.len() as isize
+        } else {
+            index
+        };
+
+        let out_of_bounds = match kind {
+            IndexFor::Lookup => index < 0 || index >= self.data.len() as isize,
+            IndexFor::Insertion => index < 0 || index > self.data.len() as isize,
+        };
```

Comment:
Strictly, Python's `list.insert` actually allows _any_ integer to be used for insertion, and just clips it to the start or end of the list respectively.  I don't know why exactly, it just does.

Reply:
Fixed in 9185dc2.